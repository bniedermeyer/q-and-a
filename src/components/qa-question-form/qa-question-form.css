:host {
  display: block;
}

.qa-post-submit-message {
  font-size: 0.75rem;
  padding: 10px;
}

#qa-question-error {
  color: red;
}

#qa-submit-question-btn {
  height: 30px;
  font-weight: bold;
  background-color: #fff;
}
