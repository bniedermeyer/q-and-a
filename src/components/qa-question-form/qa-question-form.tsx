import { Component, State, h } from '@stencil/core';
import { askQuestion } from '../../utils/data-fetching-utils';
import state from '../../store';

@Component({
  tag: 'qa-question-form',
  styleUrl: 'qa-question-form.css',
  shadow: true,
})
export class QaQuestionForm {
  /**
   * The content of the question to be submitted
   */
  @State() question: string;
  /**
   * Whether to display a submit confirmation
   */
  @State() displayConfirmation: boolean;
  /**
   * Error messages that are created while submitting a question
   */
  @State() error: string;

  async handleSubmit(e: Event): Promise<void> {
    e.preventDefault();
    if (!this.question) {
      this.error = 'Please type a question before submitting!';
      this.displayConfirmation = false;
      return;
    }
    try {
      await askQuestion(this.question);
      this.displayConfirmation = true;
    } catch (error) {
      this.error = error.message;
    }
    this.question = '';
  }

  private handleQuestionChange = e => {
    this.question = e.target.value;
    if (this.displayConfirmation) {
      this.displayConfirmation = false;
    }
    if (this.error) {
      this.error = undefined;
    }
  };

  render() {
    return (
      <form id="qa-question-form" onSubmit={e => this.handleSubmit(e)} part="form">
        <div>
          <textarea
            name="question"
            id="question"
            cols={30}
            rows={3}
            value={this.question}
            onChange={this.handleQuestionChange}
            aria-label="Submit your question here"
            part="question-textarea"
          ></textarea>
        </div>
        <button type="submit" id="qa-submit-question-btn" style={{ color: state.primaryColor, border: `3px solid ${state.primaryColor}` }} part="submit-button">
          Ask Question
        </button>
        {this.displayConfirmation ? (
          <span class="qa-post-submit-message" id="qa-question-confirm" style={{ color: state.secondaryColor }} part="confirmation-message">
            Question Asked!
          </span>
        ) : null}
        {this.error ? (
          <span class="qa-post-submit-message" id="qa-question-error" part="error-message">
            {this.error}
          </span>
        ) : null}
      </form>
    );
  }
}
