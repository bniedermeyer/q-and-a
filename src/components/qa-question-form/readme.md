# qa-question-form



<!-- Auto Generated Below -->


## Shadow Parts

| Part                     | Description |
| ------------------------ | ----------- |
| `"confirmation-message"` |             |
| `"error-message"`        |             |
| `"form"`                 |             |
| `"loading-message"`      |             |
| `"question-textarea"`    |             |
| `"submit-button"`        |             |


## Dependencies

### Used by

 - [q-and-a](../q-and-a)

### Graph
```mermaid
graph TD;
  q-and-a --> qa-question-form
  style qa-question-form fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
