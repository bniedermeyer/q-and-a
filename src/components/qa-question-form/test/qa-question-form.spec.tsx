import { newSpecPage } from '@stencil/core/testing';
import { QaQuestionForm } from '../qa-question-form';

describe('qa-question-form', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [QaQuestionForm],
      html: `<qa-question-form></qa-question-form>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
