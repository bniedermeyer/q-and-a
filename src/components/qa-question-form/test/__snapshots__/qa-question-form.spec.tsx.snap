// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`qa-question-form renders 1`] = `
<qa-question-form>
  <mock:shadow-root>
    <form id="qa-question-form">
      <div>
        <textarea aria-label="Submit your question here" cols="30" id="question" name="question" rows="3"></textarea>
      </div>
      <button id="qa-submit-question-btn" type="submit" style="color: #10915b; border: 3px solid #10915b;">
        Ask Question
      </button>
    </form>
  </mock:shadow-root>
</qa-question-form>
`;
